src/rpc/blockchain.cpp

```diff
@@ -144,6 +138,12 @@ UniValue blockToJSON(const CBlock& block, const CBlockIndex* blockindex, bool tx
 
     if (blockindex->pprev)
         result.push_back(Pair("previousblockhash", blockindex->pprev->GetBlockHash().GetHex()));
+    LOCK(cs_main);
+    int confirmations = -1;
```

Comment:
Avoid lock here by receiving the pindex tip? (with the cost of jumping thru pprev).

Reply:
That still wouldn't solve it for the chainActive.Next() call, so you'd need cs_main either way. Less code diff to just do it as-is.

Reply:
Something along this should work:
```cpp
const CBlockIndex* pindex = tip;
const CBlockIndex* pnext = nullptr;
int confirmations = 0;
while (pindex && pindex != blockindex && pindex->nHeight >= blockindex->nHeight) {
    pnext = pindex;
    pindex = pindex->pprev;
    confirmations++;
}
if (pindex != blockindex) confirmations = -1;
```
This approach avoids accessing `chainActive`, `blockToJSON` just needs to receive the tip. This sounds more correct to me because a RPC can take a snapshot of the tip (and then leave cs_main) and all the remaining execution is consistent with that tip.

Reply:
Submitted suggestion as #12151.